sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java

```diff
@@ -0,0 +1,488 @@
+ *
+ * }</pre>
+ *
+ * <h3>Writing to a JMS destination</h3>
+ *
+ * JmsIO sink supports writing test messages to a JMS destination on a broker.
+ * To configure a JMS sink, you must specify a {@link javax.jms.ConnectionFactory} and a
+ * {@link javax.jms.Destination} name.
+ * For instance:
+ *
+ * <pre>{@code
+ *
+ *  pipeline
+ *    .apply(...) // returns PCollection<String>
+ *    .apply(JmsIO.write()
+ *        .withConnectionFactory(myConnectionFactory)
+ *        .withQueue("my-queue")
+ *
+ * }</pre>
+ */
+public class JmsIO {
+
+  private static final Logger LOG = LoggerFactory.getLogger(JmsIO.class);
+
+  public static Read read() {
+    return new Read();
+  }
+
+  public static Write write() {
+    return new Write();
+  }
+
+  /**
+   * A {@link PTransform} to read from a JMS destination. See {@link JmsIO} for more
+   * information on usage and configuration.
+   */
+  public static class Read extends PTransform<PBegin, PCollection<JmsRecord>> {
+
+    public Read withConnectionFactory(ConnectionFactory connectionFactory) {
+      return new Read(connectionFactory, queue, topic, maxNumRecords, maxReadTime);
+    }
+
+    public Read withQueue(String queue) {
+      return new Read(connectionFactory, queue, topic, maxNumRecords, maxReadTime);
+    }
+
+    public Read withTopic(String topic) {
+      return new Read(connectionFactory, queue, topic, maxNumRecords, maxReadTime);
+    }
+
+    public Read withMaxNumRecords(long maxNumRecords) {
+      return new Read(connectionFactory, queue, topic, maxNumRecords, maxReadTime);
+    }
+
+    public Read withMaxReadTime(Duration maxReadTime) {
+      return new Read(connectionFactory, queue, topic, maxNumRecords, maxReadTime);
+    }
+
+    @Override
+    public PCollection<JmsRecord> apply(PBegin input) {
+      // handles unbounded source to bounded conversion if maxNumRecords is set.
+      Unbounded<JmsRecord> unbounded = org.apache.beam.sdk.io.Read.from(createSource());
+
+      PTransform<PInput, PCollection<JmsRecord>> transform = unbounded;
+
+      if (maxNumRecords < Long.MAX_VALUE) {
+        transform = unbounded.withMaxNumRecords(maxNumRecords);
+      } else if (maxReadTime != null) {
+        transform = unbounded.withMaxReadTime(maxReadTime);
+      }
+
+      return input.getPipeline().apply(transform);
+    }
+
+    @Override
+    public void validate(PBegin input) {
+      if (connectionFactory == null) {
+        throw new IllegalArgumentException("ConnectionFactory not specified");
+      }
+      if (queue == null && topic == null) {
+        throw new IllegalArgumentException("Either queue or topic not specified");
+      }
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+
+      if (queue != null) {
+        builder.add(DisplayData.item("queue", queue));
+      }
+      if (topic != null) {
+        builder.add(DisplayData.item("topic", topic));
+      }
```

Comment:
maxNumRecords? MaxReadTime?

I think the function here is called `addIfNotDefault`